File: web/lib/llmClient.mjs
// Provider-agnostic LLM client (server-side only).
//
// Works with any OpenAI-compatible chat-completions endpoint:
//   - Groq   (default)  https://api.groq.com/openai/v1
//   - OpenAI            https://api.openai.com/v1
//   - Ollama (local)    http://localhost:11434/v1
//   - OpenRouter, etc.
//
// Config via env (set in web/.env.local — NEVER commit, NEVER expose to browser):
//   LLM_PROVIDER = groq | openai | ollama | custom   (default: groq)
//   LLM_API_KEY  = ...        (or GROQ_API_KEY / OPENAI_API_KEY)
//   LLM_MODEL    = ...        (default depends on provider)
//   LLM_BASE_URL = ...        (override for custom/self-hosted)

export class LLMUnavailable extends Error {}

const PROVIDER_DEFAULTS = {
  groq: { baseUrl: "https://api.groq.com/openai/v1", model: "llama-3.3-70b-versatile", keyEnv: "GROQ_API_KEY" },
  openai: { baseUrl: "https://api.openai.com/v1", model: "gpt-4o-mini", keyEnv: "OPENAI_API_KEY" },
  ollama: { baseUrl: "http://localhost:11434/v1", model: "llama3.1", keyEnv: null },
  custom: { baseUrl: "", model: "", keyEnv: null },
};

export function getLLMConfig() {
  const provider = (process.env.LLM_PROVIDER || "groq").toLowerCase();
  const d = PROVIDER_DEFAULTS[provider] || PROVIDER_DEFAULTS.groq;
  const apiKey =
    process.env.LLM_API_KEY ||
    (d.keyEnv ? process.env[d.keyEnv] : "") ||
    "";
  return {
    provider,
    baseUrl: process.env.LLM_BASE_URL || d.baseUrl,
    model: process.env.LLM_MODEL || d.model,
    apiKey,
    // Ollama needs no key; everyone else does.
    requiresKey: provider !== "ollama",
  };
}

export function llmConfigured() {
  const c = getLLMConfig();
  return !c.requiresKey || Boolean(c.apiKey);
}

/**
 * Call the LLM and return parsed JSON. Retries once on invalid JSON.
 * Throws LLMUnavailable if no key is configured (caller should fall back).
 */
export async function callLLMJson({ system, user, temperature = 0, maxRetries = 1 }) {
  const cfg = getLLMConfig();
  if (cfg.requiresKey && !cfg.apiKey) {
    throw new LLMUnavailable(`No API key for provider "${cfg.provider}".`);
  }

  let lastErr;
  let userMsg = user;
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    const res = await fetch(`${cfg.baseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...(cfg.apiKey ? { Authorization: `Bearer ${cfg.apiKey}` } : {}),
      },
      body: JSON.stringify({
        model: cfg.model,
        temperature,
        response_format: { type: "json_object" },
        messages: [
          { role: "system", content: system },
          { role: "user", content: userMsg },
        ],
      }),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => "");
      throw new Error(`LLM HTTP ${res.status}: ${text.slice(0, 300)}`);
    }

    const data = await res.json();
    const content = data.choices?.[0]?.message?.content ?? "";
    try {
      return JSON.parse(content);
    } catch (e) {
      lastErr = e;
      // Ask the model to fix its JSON on the next attempt.
      userMsg = `${user}\n\nYour previous reply was not valid JSON. Reply with ONLY a valid JSON object, no prose.`;
    }
  }
  throw new Error(`LLM returned invalid JSON after retries: ${lastErr}`);
}

File: app/agents/llm_client.py
"""Provider-agnostic LLM client (stdlib only, no extra deps).

Calls any OpenAI-compatible /chat/completions endpoint and returns parsed JSON.
Used by the LLM agents. Raises LLMUnavailable when no key is configured so the
pipeline can fall back to rules.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request

from ..config import LLMConfig, load_llm_config


class LLMUnavailable(RuntimeError):
    """Raised when the LLM cannot be used (e.g. no API key)."""


def call_llm_json(
    system: str,
    user: str,
    *,
    temperature: float = 0.0,
    max_retries: int = 1,
    config: LLMConfig | None = None,
    timeout: float = 30.0,
) -> dict:
    """Send a chat completion and return the parsed JSON object."""
    cfg = config or load_llm_config()
    if cfg.requires_key and not cfg.api_key:
        raise LLMUnavailable(f'No API key for provider "{cfg.provider}".')

    headers = {"Content-Type": "application/json"}
    if cfg.api_key:
        headers["Authorization"] = f"Bearer {cfg.api_key}"

    user_msg = user
    last_err: Exception | None = None
    for _ in range(max_retries + 1):
        body = json.dumps(
            {
                "model": cfg.model,
                "temperature": temperature,
                "response_format": {"type": "json_object"},
                "messages": [
                    {"role": "system", "content": system},
                    {"role": "user", "content": user_msg},
                ],
            }
        ).encode("utf-8")

        req = urllib.request.Request(
            f"{cfg.base_url}/chat/completions", data=body, headers=headers, method="POST"
        )
        try:
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                data = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            detail = e.read().decode("utf-8", "ignore")[:300]
            raise RuntimeError(f"LLM HTTP {e.code}: {detail}") from e

        content = (data.get("choices") or [{}])[0].get("message", {}).get("content", "")
        try:
            return json.loads(content)
        except json.JSONDecodeError as e:
            last_err = e
            user_msg = f"{user}\n\nYour previous reply was not valid JSON. Reply with ONLY a valid JSON object."

    raise RuntimeError(f"LLM returned invalid JSON after retries: {last_err}")
